src/metaprob/generative_functions/impl.cljc

```diff
@@ -0,0 +1,10 @@
+(ns metaprob.generative-functions.impl)
+
+;; Most general way of creating a generative function: provide implementations of its
+;; methods. All other ways of creating generative functions boil down, ultimately, to
+;; a call to this function.
+(defn make-generative-function
+  ([run-in-clojure make-constrained-generator]
+   (make-generative-function run-in-clojure make-constrained-generator {}))
+  ([run-in-clojure make-constrained-generator others]
+   (with-meta run-in-clojure (assoc others :make-constrained-generator make-constrained-generator))))
```

Comment:
This is in a different namespace because it needs to be available at an earlier compilation stage than the macros that use it. Another option would be extract out this and the macros themselves into a separate macro namespace.